### PR TITLE
Fix!(hive): remove UNIQUE constraints as unsupported

### PR DESCRIPTION
**Pull request by dmoore247**
Hive/Spark/Databricks do not support unique column constraints.

### CODE REVIEW

File: sqlglot/dialects/hive.py [hunk 1]
```diff
@@ -480,6 +480,13 @@ class Generator(generator.Generator):
             exp.VolatileProperty: exp.Properties.Location.UNSUPPORTED,
         }
 
+        def create_sql(self, expression: exp.Create) -> str:
+            # remove Unique Column Constraints
+            for constraint in expression.find_all(exp.UniqueColumnConstraint):
+                if constraint.parent:
+                    constraint.parent.pop()
+            return super().create_sql(expression)
```
Comment (GeorgeSittas) [inline sqlglot/dialects/hive.py:488]: This looks iffy, we already have an entry for `exp.Create` in `TRANSFORMS`.. I'd rather remove that entry and have this method return `create_with_partitions_sql(...)` instead.

**Review by GeorgeSittas — COMMENTED** [review]
I tried some DDLs in Spark to check this, and it seems like `UNIQUE` is rejected. Please, for future PRs / issues link any necessary docs / sources because this fact-checking is time-consuming.

Thanks for the PR in any case.

File: sqlglot/dialects/hive.py (hunk 1 above)
Comment (GeorgeSittas) [reply sqlglot/dialects/hive.py:488]: Spark2 also overrides `create_sql` to return `create_with_partitions_sql`, should we instead fall back to Hive's `create_sql` method?

Comment (dmoore247) [reply sqlglot/dialects/hive.py:488]: Ok, makes sense to return `create_with_partitions_sql`

**Comment by dmoore247** [comment]
@GeorgeSittas it's a challenge to find docs on features that are not supported. :)

Would including errors and error messages be useful documentation for PRs?
Would opening issues here to document be helpful, or just include with the PR?

**Comment by GeorgeSittas** [comment]
> @GeorgeSittas it's a challenge to find docs on features that are not supported. :)
> 
> Would including errors and error messages be useful documentation for PRs? Would opening issues here to document be helpful, or just include with the PR?

Yep, any testing / errors encountered in the corresponding environments definitely help if there are no available references to check that a feature is unsupported. I get where you're coming from, it's just that we're going to double-check these things anyway, so any additional info is super helpful.

File: sqlglot/dialects/hive.py (hunk 1 above)
Comment (dmoore247) [reply sqlglot/dialects/hive.py:488]: My first attempt to return `create_with_partitions_sql(...)` results in an infinite loop. I'll look at this more closely later in the day, though at first glance looks like `generator:create_with_partitions_sql(...)` calls `create_sql`

File: sqlglot/dialects/hive.py [hunk 2]
```diff
@@ -51,6 +51,13 @@
 DIFF_MONTH_SWITCH = ("YEAR", "QUARTER", "MONTH")
 
 
+def _create_sql(self, expression: exp.Create) -> str:
+    # remove Unique Column Constraints
+    for constraint in expression.find_all(exp.UniqueColumnConstraint):
+        if constraint.parent:
+            constraint.parent.pop()
+    return create_with_partitions_sql(self, expression)
```
Comment (GeorgeSittas) [inline sqlglot/dialects/hive.py:59]: Can we just move this method inside Hive's `Generator` class? Also, can we merge its logic with Spark2's `_create_sql`? If that's possible then Spark will pick that up through inheritance.

File: sqlglot/dialects/hive.py [hunk 3]
```diff
@@ -429,7 +436,8 @@ class Generator(generator.Generator):
             if e.args.get("allow_null")
             else "NOT NULL",
             exp.VarMap: var_map_sql,
-            exp.Create: create_with_partitions_sql,
+            #exp.Create: create_with_partitions_sql,
+            exp.Create: _create_sql,
```
Comment (GeorgeSittas) [inline sqlglot/dialects/hive.py:440]: This won't be needed if `_create_sql` becomes a `Generator` method.

**Review by GeorgeSittas — COMMENTED** [review]
Looks good, couple of comments. Need to run `make style` but looks ok I think.

File: sqlglot/dialects/hive.py (hunk 2 above)
Comment (dmoore247) [reply sqlglot/dialects/hive.py:59]: @GeorgeSittas I wanted to ask you this same question...
[Hive supports CREATE TEMPORARY TABLE AS SELECT](https://docs.cloudera.com/HDPDocuments/HDP3/HDP-3.0.1/using-hiveql/content/hive_create_a_hive_temporary_table.html) ... so moving the spark2 temporary table logic to hive would undo that.

I would be fine with moving the UNIQUE logic to spark2 and not worrying about that for Hive at this time. Thoughts?

Comment (GeorgeSittas) [reply sqlglot/dialects/hive.py:59]: Ok no problem leaving Spark2 `_create_sql` as is then.

> I would be fine with moving the UNIQUE logic to spark2 and not worrying about that for Hive at this time. Thoughts?

Nah let's just keep it in Hive. Can we instead return `super(Hive.Generator, self).create_sql(expression)` for Spark2 in the fallback case where we don't find a temporary property? That way we'll reuse Hive's logic, but we need to first move `_create_sql` in Hive's generator and remove the `_` prefix to associate it with `exp.Create`.

File: sqlglot/dialects/hive.py (hunk 3 above)
Comment (dmoore247) [reply sqlglot/dialects/hive.py]: To override `create_sql` in Hive Generator class, the execution flow will go 
```
hive.py:create_sql() -> return create_with_partitions_sql ->
dialect.py:create_with_partitions_sql -> self.create_sql() ->
hive.py:create_sql() -> ...
```

Comment (GeorgeSittas) [reply sqlglot/dialects/hive.py]: Hm I see, makes sense. Can probably leave it as is then

File: sqlglot/dialects/hive.py [hunk 4]
```diff
@@ -51,6 +51,14 @@
 DIFF_MONTH_SWITCH = ("YEAR", "QUARTER", "MONTH")
 
 
+def _create_sql(self, expression: exp.Create) -> str:
+    # remove Unique Column Constraints
+    for constraint in expression.find_all(exp.UniqueColumnConstraint):
+        if constraint.parent:
+            constraint.parent.pop()
```
Comment (GeorgeSittas) [inline sqlglot/dialects/hive.py:58]: We need to copy `expression` before popping here because this is a mutation

**Comment by dmoore247** [comment]
@GeorgeSittas 
This creates an infinite loop
```python
hive.py:
        def create_sql(self, expression: exp.Create) -> str:
            # remove Unique Column Constraints
            expression = expression.copy()
            for constraint in expression.find_all(exp.UniqueColumnConstraint):
                if constraint.parent:
                    constraint.parent.pop()
            return create_with_partitions_sql(self, expression)
```

**Comment by GeorgeSittas** [comment]
> @GeorgeSittas This creates an infinite loop
> 
> ```python
> hive.py:
>         def create_sql(self, expression: exp.Create) -> str:
>             # remove Unique Column Constraints
>             expression = expression.copy()
>             for constraint in expression.find_all(exp.UniqueColumnConstraint):
>                 if constraint.parent:
>                     constraint.parent.pop()
>             return create_with_partitions_sql(self, expression)
> ```

Yeah because `create_with_partitions_sql` ends up returning `self.create_sql(expression)` like you mentioned in an earlier followup. Maybe the way you have it now is ok. What's causing these CI failures?

**Comment by dmoore247** [comment]
I'm working on the CI features... my earlier PR 'USING PARQUET` is messing with this PR. I'll need a little bit of time to fix.

**Comment by dmoore247** [comment]
We had overrides on _create_sql in hive,spark2,spark and databricks. This last change consolidated all of them into a hive.py:_create_sql. There is one Generator method to add temp table storage provider that was unavoidable ([Hive does not need a storage provider ](https://docs.cloudera.com/HDPDocuments/HDP3/HDP-3.0.1/using-hiveql/content/hive_create_a_hive_temporary_table.html))

File: sqlglot/dialects/hive.py [hunk 5]
```diff
@@ -480,6 +506,10 @@ class Generator(generator.Generator):
             exp.VolatileProperty: exp.Properties.Location.UNSUPPORTED,
         }
 
+        def temporary_storage_provider(self, expression):
+            # Hive has no temporary storage provider (there are hive settings though)
+            return expression
```
Comment (GeorgeSittas) [inline sqlglot/dialects/hive.py:511]: Let's add type annotations here

File: sqlglot/dialects/spark2.py [hunk 6]
```diff
@@ -236,6 +218,13 @@ class Generator(Hive.Generator):
         WRAP_DERIVED_VALUES = False
         CREATE_FUNCTION_RETURN_AS = False
 
+        def temporary_storage_provider(self, expression):
```
Comment (GeorgeSittas) [inline sqlglot/dialects/spark2.py:221]: Ditto

**Review by GeorgeSittas — APPROVED** [review]

**Comment by GeorgeSittas** [comment]
Thanks for the quick iterations on this, @dmoore247!

File: sqlglot/dialects/spark2.py (hunk 6 above)
Comment (dmoore247) [reply sqlglot/dialects/spark2.py:221]: of course, my bad.

File: sqlglot/dialects/hive.py [hunk 7]
```diff
@@ -480,6 +506,10 @@ class Generator(generator.Generator):
             exp.VolatileProperty: exp.Properties.Location.UNSUPPORTED,
         }
 
+        def temporary_storage_provider(self, expression: exp.Create):
```
Comment (GeorgeSittas) [inline sqlglot/dialects/hive.py:509]:
```suggestion
        def temporary_storage_provider(self, expression: exp.Create) -> exp.Create:
```

File: sqlglot/dialects/spark2.py [hunk 8]
```diff
@@ -236,6 +218,13 @@ class Generator(Hive.Generator):
         WRAP_DERIVED_VALUES = False
         CREATE_FUNCTION_RETURN_AS = False
 
+        def temporary_storage_provider(self, expression: exp.Create):
```
Comment (GeorgeSittas) [inline sqlglot/dialects/spark2.py:221]:
```suggestion
        def temporary_storage_provider(self, expression: exp.Create) -> exp.Create:
```

Comment (dmoore247) [reply sqlglot/dialects/spark2.py]: Thanks!

**Pull request merged**